### PR TITLE
[ITSM][ServiceNow] Implement ServiceNow Controller

### PR DESCRIPTION
**Pull request by sidmore**
<!--- This repository only accepts pull requests related to open issues, please link the open issue in description below. See https://help.github.com/articles/closing-issues-using-keywords/ to learn about automation. 
ITSMSkill needs to be able to process ServiceNow Events this PR enables a Controller for the same.

### Purpose
*What is the context of this pull request? Why is it being done?*
Implement ServiceNow API controller to receive callback and create EventActivity

### Changes
*Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)*
a. Add ServiceNowController
b. Add ServiceNowMessageReceiver
c. Update DefaultActivityHandler to process Proactive Event and update conversation

### Tests
*Is this covered by existing tests or new ones? If no, why not?*
a. Added UT for ServiceNowController
b. Need to E2E notification Test

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*
None

#### General
- [x] I have commented my code, particularly in hard-to-understand areas	
- [x] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

### CODE REVIEW

File: skills/csharp/experimental/itsmskill/Proactive/ServiceNowProactiveState.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿namespace ITSMSkill.Proactive
+{
+    using Microsoft.Bot.Builder;
+
+    public class ServiceNowProactiveState : BotState
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Proactive/ServiceNowProactiveState.cs]: unused?

File: skills/csharp/experimental/itsmskill/Extensions/NotificationToAdaptiveCardExtension.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    using ITSMSkill.Models.ServiceNow;
+
+    public static class NotificationToAdaptiveCardExtension
+    {
+        /// <summary>
+        /// Creates AdaptiveCard from Notification.
+        /// </summary>
+        /// <param name="serviceNowNotification">ServiceNow Notification.</param>
+        /// <returns>AdaptiveCard.</returns>
+        public static AdaptiveCard ToAdaptiveCard(this ServiceNowNotification serviceNowNotification)
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Extensions/NotificationToAdaptiveCardExtension.cs]: use json with LocaleTemplateManager?

File: skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs [hunk 3]
```diff
@@ -53,14 +67,91 @@ protected override Task OnTeamsSigninVerifyStateAsync(ITurnContext<IInvokeActivi
...
+        /// Continue the conversation callback.
+        /// </summary>
+        /// <param name="context">Turn context.</param>
+        /// <param name="message">Activity text.</param>
+        /// <returns>Bot Callback Handler.</returns>
+        private BotCallbackHandler ContinueConversationCallback(ITurnContext context, ServiceNowNotification notification)
+        {
+            return async (turnContext, cancellationToken) =>
+            {
+                var activity = context.Activity.CreateReply();
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs]: Recommend to use LocaleTemplateManager for generating activity.

File: skills/csharp/experimental/itsmskill/Models/ServiceNow/ServiceNowNotification.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿using Newtonsoft.Json;
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Models/ServiceNow/ServiceNowNotification.cs]: License headers etc.

I am wondering what's the purpose of this update? 
For proactive demo, it has already been in enterprise VA.
If for ServiceNow, is it a standard structure? Where could I find documents about this? Thanks

File: skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs [hunk 5]
```diff
@@ -53,14 +67,91 @@ protected override Task OnTeamsSigninVerifyStateAsync(ITurnContext<IInvokeActivi
...
+                        // TODO: Implement a proactive subscription manager for mapping Notification to ConversationReference
+                        var conversationReference = proactiveModel["Key"].Conversation;
+
+                        await turnContext.Adapter.ContinueConversationAsync(_appCredentials.MicrosoftAppId, conversationReference, ContinueConversationCallback(turnContext, eventData), cancellationToken);
+                        break;
+                    }
+
+                default:
+                    {
+                        await turnContext.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"Unknown Event '{ev.Name ?? "undefined"}' was received but not processed."));
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs]: We should always call ` _dialog.RunAsync(turnContext, _dialogStateAccessor, cancellationToken);` for other events.
No need to handle them (like TokenEvents.TokenResponseEventName) here.

File: skills/csharp/experimental/itsmskill/Controllers/ServiceNow/ServiceNowControllerBase.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace ITSMSkill.Controllers.ServiceNow
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+    using ITSMSkill.Models.ServiceNow;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Bot.Builder;
+
+    public abstract class ServiceNowControllerBase : Controller
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Controllers/ServiceNow/ServiceNowControllerBase.cs]: I don't find it is necessary to use ServiceNowControllerBase and ServiceNowController. We could just DeserializeObject and call Receive in Controller. One could even wrap it to a template function.

**Comment by xieofxie** [comment]
@okulygin @ryanlengel @lzc850612 https://github.com/microsoft/botframework-solutions/pull/3232 is here now

File: skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs (hunk 5 above)
Comment (sidmore) [reply skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs]: @xieofxie thanks added that in default

File: skills/csharp/experimental/itsmskill/Controllers/ServiceNow/ServiceNowController.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Bot.Builder;
+    using Microsoft.Bot.Builder.Integration.AspNet.Core;
+    using Microsoft.Bot.Schema;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Controller to Process ServiceNow Incident Events
+    /// </summary>
+    [Route("api/servicenow/processIncidents")]
```
Comment (lzc850612) [inline skills/csharp/experimental/itsmskill/Controllers/ServiceNow/ServiceNowController.cs]: api/servicenow/incidents would be a more REST name

File: skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs [hunk 8]
```diff
@@ -60,14 +74,91 @@ protected override Task OnTeamsSigninVerifyStateAsync(ITurnContext<IInvokeActivi
...
-        protected override Task OnEventActivityAsync(ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
+        protected override async Task OnEventActivityAsync(ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
         {
-            return _dialog.RunAsync(turnContext, _dialogStateAccessor, cancellationToken);
+            var ev = turnContext.Activity.AsEventActivity();
+            var value = ev.Value?.ToString();
+
+            switch (ev.Name)
+            {
+                case TokenEvents.TokenResponseEventName:
```
Comment (xieofxie) [inline skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs]: It is same as default now

Comment (sidmore) [reply skills/csharp/experimental/itsmskill/Bots/DefaultActivityHandler.cs:84]: removed

**Comment by xieofxie** [comment]
Hi @sidmore It is ready to merge now, please merge the branch yourself.
In case you have some updates/fixes, approvers will not merge the pr.
Thanks

**Pull request merged**